### PR TITLE
clear mpicc and friends before each build

### LINKED ISSUE #17322: Installation issue: openmpi@3.1.6 %gcc@8.4.0
### Steps to reproduce the issue

```console
$ spack install openmpi@3.1.6 %gcc@8.4.0 +cuda+cxx+cxx_exceptions fabrics=ucx -gpfs+pmi schedulers=slurm
```

### Information on your system

* **Spack:** 0.14.2-1624-212299a
* **Python:** 3.6.8
* **Platform:** linux-centos7-skylake_avx512

packages.yaml:
```console
packages:
  binutils:
    version: [2.29.1]
  netcdf:
    variants: maxdims=65536 maxvars=524288
  hdf5:
    variants: +cxx+hl+fortran+szip
  slurm:
    paths:
      slurm@19-05-0-2: /nopt/slurm/current
    buildable: False
  intel-mkl:
     paths:
       intel-mkl@2020.1.217: /nopt/nrel/apps/compilers/intel/2020.1.217/mkl
     buildable: False
  rdma-core:
     paths:
       rdma-core: /lib64
     buildable: False
  hwloc:
    variants: +cuda ~nvml ~gl
  fftw:
    variants: +fma simd=avx2,avx512
  all:
    compiler: [gcc@8.4.0, intel@19.1.1.217]
    # Spack uses build_type=RelWithDebInfo by default for CMake packages
    # and creates huge binaries unless you do this:
    variants: build_type=Release
    target: [x86_64]
    providers:
      mpi: [openmpi@3.1.3, intel-mpi@2019.5.222]
      blas: [netlib-lapack@3.8.0, intel-mkl@2020.1.217]
      lapack: [netlib-lapack@3.8.0, intel-mkl@2020.1.217]
```

### Additional information
* The build fails during UCX installation.
* I'm also confused why UCX is being built without CUDA, since OpenMPI is being requested with it.
* Not sure why shmem.h is being included from the OpenMPI 1.10.7 build done prior to this one (openmpi@3.1.6).

```console
...
==> 9016: Installing ucx
==> Using cached archive: /nopt/nrel/apps/base/2020-05-12/spack/var/spack/cache/_source-cache/archive/e4/e400f7aa5354971c8f5ac6b881dc2846143851df868088c37d432c076445628d.tar.gz
==> Staging archive: /tmp/cchang/spack-stage/spack-stage-ucx-1.8.0-wix255r63ctwidsxwkgn74r3cttzmmde/ucx-1.8.0.tar.gz
==> Created stage in /tmp/cchang/spack-stage/spack-stage-ucx-1.8.0-wix255r63ctwidsxwkgn74r3cttzmmde
==> No patches needed for ucx
==> 9016: ucx: Building ucx [AutotoolsPackage]
==> 9016: ucx: Executing phase: 'autoreconf'
==> 9016: ucx: Executing phase: 'configure'
==> 9016: ucx: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16'

4 errors found in build log:
     826    Making all in test/mpi
     827    make[2]: Entering directory `/tmp/cchang/spack-stage/spack-stage-ucx-1.8.0-wix255r63ctwidsxwkgn74r3cttzmmde/spack-src/test/mpi'
     828      CC       test_memhooks.o
     829      CC       shmem_pingpong.o
     830      CC       test_memhooks_lib.lo
     831    shmem_pingpong.c: In function 'run_pingpong':
  >> 832    shmem_pingpong.c:72:34: error: passing argument 1 of 'shmem_int_wait_until' discards 'volatile' qualifier from pointer target type [-Werror=disc
            arded-qualifiers]
     833                 shmem_int_wait_until(rsn, SHMEM_CMP_EQ, sn);
     834                                      ^~~
     835    In file included from shmem_pingpong.c:13:
     836    /nopt/nrel/apps/base/2020-05-12/spack/opt/spack/linux-centos7-x86_64/gcc-8.4.0/openmpi-1.10.7-unlnregriwszfdzpmhbspyijylt4vras/include/shmem.h:2
            88:49: note: expected 'int *' but argument is of type 'volatile int *'
     837     OSHMEM_DECLSPEC  void shmem_int_wait_until(int *addr, int cmp, int value);
     838                                                ~~~~~^~~~
     839    cc1: all warnings being treated as errors
  >> 840    make[2]: *** [shmem_pingpong.o] Error 1
     841    make[2]: *** Waiting for unfinished jobs....
     842    make[2]: Leaving directory `/tmp/cchang/spack-stage/spack-stage-ucx-1.8.0-wix255r63ctwidsxwkgn74r3cttzmmde/spack-src/test/mpi'
  >> 843    make[1]: *** [all-recursive] Error 1
     844    make[1]: Leaving directory `/tmp/cchang/spack-stage/spack-stage-ucx-1.8.0-wix255r63ctwidsxwkgn74r3cttzmmde/spack-src'
  >> 845    make: *** [all] Error 2

See build log for details:
  /tmp/cchang/spack-stage/spack-stage-ucx-1.8.0-wix255r63ctwidsxwkgn74r3cttzmmde/spack-build-out.txt

==> Error: Failed to install ucx due to ChildError: ProcessError: Command exited with status 2:
    'make' '-j16'
4 errors found in build log:
     826    Making all in test/mpi
     827    make[2]: Entering directory `/tmp/cchang/spack-stage/spack-stage-ucx-1.8.0-wix255r63ctwidsxwkgn74r3cttzmmde/spack-src/test/mpi'
     828      CC       test_memhooks.o
     829      CC       shmem_pingpong.o
     830      CC       test_memhooks_lib.lo
     831    shmem_pingpong.c: In function 'run_pingpong':
  >> 832    shmem_pingpong.c:72:34: error: passing argument 1 of 'shmem_int_wait_until' discards 'volatile' qualifier from pointer target type [-Werror=disc
            arded-qualifiers]
     833                 shmem_int_wait_until(rsn, SHMEM_CMP_EQ, sn);
     834                                      ^~~
     835    In file included from shmem_pingpong.c:13:
     836    /nopt/nrel/apps/base/2020-05-12/spack/opt/spack/linux-centos7-x86_64/gcc-8.4.0/openmpi-1.10.7-unlnregriwszfdzpmhbspyijylt4vras/include/shmem.h:2
            88:49: note: expected 'int *' but argument is of type 'volatile int *'
     837     OSHMEM_DECLSPEC  void shmem_int_wait_until(int *addr, int cmp, int value);
     838                                                ~~~~~^~~~
     839    cc1: all warnings being treated as errors
  >> 840    make[2]: *** [shmem_pingpong.o] Error 1
     841    make[2]: *** Waiting for unfinished jobs....
     842    make[2]: Leaving directory `/tmp/cchang/spack-stage/spack-stage-ucx-1.8.0-wix255r63ctwidsxwkgn74r3cttzmmde/spack-src/test/mpi'
  >> 843    make[1]: *** [all-recursive] Error 1
     844    make[1]: Leaving directory `/tmp/cchang/spack-stage/spack-stage-ucx-1.8.0-wix255r63ctwidsxwkgn74r3cttzmmde/spack-src'
  >> 845    make: *** [all] Error 2

See build log for details:
  /tmp/cchang/spack-stage/spack-stage-ucx-1.8.0-wix255r63ctwidsxwkgn74r3cttzmmde/spack-build-out.txt
Traceback (most recent call last):
  File "/nopt/nrel/apps/base/2020-05-12/spack/lib/spack/spack/build_environment.py", line 838, in child_process
    return_value = function()
  File "/nopt/nrel/apps/base/2020-05-12/spack/lib/spack/spack/installer.py", line 1140, in build_process
    phase(pkg.spec, pkg.prefix)
  File "/nopt/nrel/apps/base/2020-05-12/spack/lib/spack/spack/package.py", line 108, in phase_wrapper
    phase(spec, prefix)
  File "/nopt/nrel/apps/base/2020-05-12/spack/lib/spack/spack/build_systems/autotools.py", line 331, in build
    inspect.getmodule(self).make(*self.build_targets)
  File "/nopt/nrel/apps/base/2020-05-12/spack/lib/spack/spack/build_environment.py", line 133, in __call__
    return super(MakeExecutable, self).__call__(*args, **kwargs)
  File "/nopt/nrel/apps/base/2020-05-12/spack/lib/spack/spack/util/executable.py", line 189, in __call__
    proc.returncode, long_msg)
spack.util.executable.ProcessError: Command exited with status 2:
    'make' '-j16'

==> Warning: Skipping build of openmpi since ucx failed
==> Error: Installation of openmpi failed.  Review log for details
```

* [spack-build-out.txt](https://github.com/spack/spack/files/4855076/spack-build-out.txt)
* [spack-build-env.txt](https://github.com/spack/spack/files/4855078/spack-build-env.txt)
* [spack-configure-args.txt](https://github.com/spack/spack/files/4855093/spack-configure-args.txt)

@hppritcha

### General information

- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have run `spack maintainers <name-of-the-package>` and @mentioned any maintainers
- [x] I have uploaded the build log and environment files
- [x] I have searched the issues of this repo and believe this is not a duplicate

### PR DESCRIPTION
**Pull request by s-sajid-ali**
Fixes https://github.com/spack/spack/issues/17322

@becker33

### CODE REVIEW

File: lib/spack/spack/build_environment.py [hunk 1]
```diff
@@ -174,6 +174,14 @@ def clean_environment():
     for v in build_system_vars:
         env.unset(v)
 
+    # Unset mpi environment vars. These flags should only be set by
+    # mpi providers for packages with mpi dependencies
+    mpi_vars = [
+        'MPICC', 'MPICXX', 'MPIF70', 'MPIF90'
```
Comment (becker33) [inline lib/spack/spack/build_environment.py]:
```suggestion
        'MPICC', 'MPICXX', 'MPIFC', 'MPIF77', 'MPIF90'
```

**Review by becker33 — CHANGES_REQUESTED** [review]
One quick change to the env names needed, otherwise looks good.

File: lib/spack/spack/build_environment.py [hunk 2]
```diff
@@ -174,6 +174,14 @@ def clean_environment():
     for v in build_system_vars:
         env.unset(v)
 
+    # Unset mpi environment vars. These flags should only be set by
+    # mpi providers for packages with mpi dependencies
+    mpi_vars = [
+        'MPICC', 'MPICXX', 'MPIFC', 'MPIF70', 'MPIF90'
```
Comment (becker33) [inline lib/spack/spack/build_environment.py]:
```suggestion
        'MPICC', 'MPICXX', 'MPIFC', 'MPIF77', 'MPIF90'
```

**Review by becker33 — CHANGES_REQUESTED** [review]
You addressed one of the two changes in the previous suggestion. MPIF77 is the name of the variable, not MPIF70.

**Review by becker33 — APPROVED** [review]

File: lib/spack/spack/build_environment.py (hunk 2 above)
Comment (s-sajid-ali) [reply lib/spack/spack/build_environment.py]: Apologies for resolving the prior thread while overlooking that suggestion.

**Comment by becker33** [comment]
@scheibelp do you want this in 0.15.1 or 0.15.2?

**Pull request merged**